### PR TITLE
[refactor] pipe: move async-specific code out of MultiProcessPipeline

### PR DESCRIPTION
**Pull request by msbaines**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/nn/pipe/async_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+
+class AsyncPipeline(MultiProcessPipeline):
+    def run(self, training: bool, batches: List[Batch], event: Optional[Event]) -> None:
+
+        """Runs pipeline parallelism.
+
+        It modifies the given batches in place.
+
+        """
```
Comment (anj-s) [inline fairscale/nn/pipe/async_pipeline.py:25]: nit: Can we update this comment to make this more descriptive of what AsyncPipeline does?

File: fairscale/nn/pipe/async_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            logging.debug(f"{torch.distributed.get_rank()}: exited event tail")
+        else:
+            logging.debug(f"{torch.distributed.get_rank()}: entered event loop")
+            event_loop.event_loop(len(batches), skip_trackers)
+            logging.debug(f"{torch.distributed.get_rank()}: exited event loop")
+
+        self.callcount += 1
+
+    def back_helper(self, output: List[Batch]) -> None:
+        pass
```
Comment (anj-s) [inline fairscale/nn/pipe/async_pipeline.py:48]: Should we add a comment about why we don't need to implement this?

File: fairscale/nn/pipe/async_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        elif self.final_stage:
+            logging.debug(f"{torch.distributed.get_rank()}: entered event tail")
+            event_loop.event_loop_tail(batches, skip_trackers)
+            logging.debug(f"{torch.distributed.get_rank()}: exited event tail")
+        else:
+            logging.debug(f"{torch.distributed.get_rank()}: entered event loop")
+            event_loop.event_loop(len(batches), skip_trackers)
+            logging.debug(f"{torch.distributed.get_rank()}: exited event loop")
+
+        self.callcount += 1
```
Comment (anj-s) [inline fairscale/nn/pipe/async_pipeline.py:45]: Just curious, do we know where this is used?

**Review by anj-s — APPROVED** [review]

**Review by sidgoyal78 — APPROVED** [review]
Thanks for the PR.

File: fairscale/nn/pipe/async_pipeline.py (hunk 3 above)
Comment (msbaines) [reply fairscale/nn/pipe/async_pipeline.py:45]: I doesn't appear to be used anywhere. Might have been added for debug. I'll remove it.

**Pull request merged**